jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/DoSHandler.java

```diff
@@ -0,0 +1,792 @@
+    {
+        // Reject if we have too many Trackers
+        if (_maxTrackers > 0 && _trackers.size() >= _maxTrackers)
+        {
+            // Try shrinking the tracker pool
+            long now = NanoTime.now();
+            _trackers.values().removeIf(tracker -> tracker.isIdle(now));
+            if (_trackers.size() >= _maxTrackers)
+            {
+                // Try shrinking the tracker pool as if we are at the next sample period already
+                long nextIdleCheck = NanoTime.now() + _trackerFactory.getSamplePeriod().getNano();
+                _trackers.values().removeIf(tracker -> tracker.isIdle(nextIdleCheck));
+                if (_trackers.size() >= _maxTrackers)
+                    return _rejectHandler.handle(request, response, callback);
+            }
+        }
+
+        // Calculate an id for the request (which may be global empty string)
+        String id = _getId.apply(request);
+
+        if (id == null)
+            return _rejectHandler.handle(request, response, callback);
+
+        // Obtain a tracker, creating a new one if necessary.  Trackers are removed if CyclicTimeouts#onExpired returns true
+        Tracker tracker = _trackers.computeIfAbsent(id, this::newTracker);
+
+        // If we are not over-limit then handle normally
+        if (tracker.onRequest(request.getBeginNanoTime()))
+            return nextHandler(request, response, callback);
+
+        // Otherwise reject the request
+        return _rejectHandler.handle(request, response, callback);
+    }
+
+    Tracker newTracker(String id)
+    {
+        Tracker tracker = _trackerFactory.newTracker(id);
+        if (_cyclicTimeouts != null)
+            _cyclicTimeouts.schedule(tracker);
+        return tracker;
+    }
+
+    @Override
+    protected void doStart() throws Exception
+    {
+        _cyclicTimeouts = new CyclicTimeouts<>(getServer().getScheduler())
+        {
+            @Override
+            protected Iterator<Tracker> iterator()
+            {
+                return _trackers.values().iterator();
+            }
+
+            @Override
+            protected boolean onExpired(Tracker tracker)
+            {
+                return tracker.isIdle(NanoTime.now());
+            }
+        };
+        addBean(_cyclicTimeouts);
+        super.doStart();
+    }
+
+    @Override
+    protected void doStop() throws Exception
+    {
+        super.doStop();
+        removeBean(_cyclicTimeouts);
+        _cyclicTimeouts.destroy();
+        _cyclicTimeouts = null;
+    }
+
+    /**
+     * A RateTracker is associated with an id, and stores request rate data.
+     */
+    public abstract static class Tracker implements CyclicTimeouts.Expirable
+    {
+        private final String _id;
+
+        protected Tracker(String id)
+        {
+            _id = id;
+        }
+
+        public String getId()
+        {
+            return _id;
+        }
+
+        public abstract boolean isIdle(long now);
```

Comment:
But it is already `Expirable`, why this method?

Reply:
So that we can find less busy trackers to remove when we hit max trackers